cmd/zpool/zpool_main.c

```diff
@@ -8688,6 +8704,122 @@ zpool_do_trim(int argc, char **argv)
 	return (error);
 }
 
+typedef struct {
+	pool_condense_func_t func;
+	pool_condense_type_t type;
+} condense_cb_t;
+
+static int
+condense_cb(zpool_handle_t *zhp, void *data)
+{
+	condense_cb_t *cb = data;
+	return (zpool_condense(zhp, cb->func, cb->type));
+}
+
+/*
+ * zpool condense -t <target> [-c | -w] <pool>
+ *
+ *	-t <target>	What to condense.
+ *	-c		Cancel. Ends any in-progress condense.
+ *	-w		Wait. Blocks until condense has completed.
+ *
+ * Condense (flush) the log spacemap on the specified pool(s).
+ */
+static int
+zpool_do_condense(int argc, char **argv)
+{
+	struct option long_options[] = {
+		{"target",	required_argument,	NULL,	't'},
+		{"cancel",	no_argument,		NULL,	'c'},
+		{"wait",	no_argument,		NULL,	'w'},
+		{0, 0, 0, 0}
+	};
+
+	struct target_map {
+		const char *name;
+		pool_condense_type_t type;
+	} targets[] = {
+		{"log-spacemap",	POOL_CONDENSE_LOG_SPACEMAP},
+		{0, 0}
+	};
+
+	condense_cb_t cb = {
+		.func = POOL_CONDENSE_START,
+		.type = POOL_CONDENSE_TYPES,
+	};
+	boolean_t wait = B_FALSE;
+
+	int c;
+	while ((c = getopt_long(argc, argv, "t:cw", long_options, NULL))
+	    != -1) {
+		switch (c) {
+		case 't': {
+			struct target_map *t;
+			for (t = targets; t->name != NULL; t++) {
+				if (strcmp(t->name, optarg) == 0) {
+					cb.type = t->type;
+					break;
+				}
+			}
+			if (t->name == NULL) {
+				(void) fprintf(stderr,
+				    gettext("invalid condense target '%s'\n"),
+				    optarg);
+				usage(B_FALSE);
+			}
+			break;
+		}
+		case 'c':
+			cb.func = POOL_CONDENSE_CANCEL;
+			break;
+		case 'w':
+			wait = B_TRUE;
+			break;
+		case '?':
+			if (optopt != 0) {
+				(void) fprintf(stderr,
+				    gettext("invalid option '%c'\n"), optopt);
+			} else {
+				(void) fprintf(stderr,
+				    gettext("invalid option '%s'\n"),
+				    argv[optind - 1]);
+			}
+			usage(B_FALSE);
+		}
+	}
+
+	if (cb.type == POOL_CONDENSE_TYPES) {
+		(void) fprintf(stderr, gettext("missing condense target\n"));
+		usage(B_FALSE);
+	}
+
+	argc -= optind;
+	argv += optind;
+
+	if (argc < 1) {
+		(void) fprintf(stderr, gettext("missing pool name argument\n"));
```

Comment:
Yeah, I didn't want an "all pools" mode here, because I don't know what might be there in the future, and it might mean different things for different pools. Same way `zpool scrub` requires a pool arg.